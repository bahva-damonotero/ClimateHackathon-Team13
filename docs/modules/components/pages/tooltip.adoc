= Tooltip

A tooltip is a short descriptive message that appears when a user hovers or focuses on an element.

== Usage Examples

=== Default

[tabs]
====
Preview::
+
--

++++
<iframe
  src="/uip/docs/examples-react/#/tooltip-example"
  width="100%"
  frameBorder="0"
  onload="resizeIframe(this)"
></iframe>
++++

--
JSX::
+
--

[source,jsx]
----
include::example$tooltip/TooltipExample.jsx[]
----

--
====

'''

=== Icon

[tabs]
====
Preview::
+
--

++++
<iframe
  src="/uip/docs/examples-react/#/tooltip-icon-example"
  width="100%"
  frameBorder="0"
  onload="resizeIframe(this)"
></iframe>
++++

--
JSX::
+
--

[source,jsx]
----
include::example$tooltip/TooltipIconExample.jsx[]
----

--
====

== Props
[grid="rows"]
|===
| *Name* | *Type* | *Default* | *Description* 
include::shared/shared.adoc[tag=className]
| content | `String` | | The content of the tooltip. 
| position | `'top'` +
\| `'left'` +
\| `'bottom'` +
\| `'right'` | `'top'` | The position of the tooltip. 
|===
