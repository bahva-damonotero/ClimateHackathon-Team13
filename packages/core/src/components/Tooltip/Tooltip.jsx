import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import tooltip from '@uswds/uswds/js/usa-tooltip';

import {
  createClassString,
  flattenChildren
} from '@boozallen-uip/uswds-react-utils';

export default function Tooltip({ children, className, content, position }) {
  const containerRef = useRef();

  useEffect(() => {
    const containerEl = containerRef.current;
    if (!containerEl) return;

    // call USWDS's `on` function to initialize the component
    tooltip.on(containerEl);

    // on dismount, call USWDS's `off` function to cleanup the component
    return () => {
      tooltip.off(containerEl);
    };
  }, []);

  return (
    <div
      className={createClassString(['bah-uswds-tooltip', className])}
      ref={containerRef}
    >
      {flattenChildren(React.Children.toArray(children)).map((child) =>
        React.cloneElement(child, {
          className: 'usa-tooltip',
          'data-position': position,
          title: content
        })
      )}
    </div>
  );
}

Tooltip.propTypes = {
  /** The content that when moused over or when focused should display the tooltip */
  children: PropTypes.node.isRequired,
  /** Any additional classes to apply */
  className: PropTypes.string,
  /** The content of the tooltip */
  content: PropTypes.string.isRequired,
  /** The position of the tooltip */
  position: PropTypes.oneOf(['top', 'left', 'bottom', 'right'])
};

Tooltip.defaultProps = {
  position: 'right'
};
