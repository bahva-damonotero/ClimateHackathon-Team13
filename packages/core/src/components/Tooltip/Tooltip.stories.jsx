/* eslint-disable react/prop-types */
import React from 'react';

import { Button } from '..';
import Tooltip from '.';

export default {
  title: 'Tooltip',
  component: Tooltip,
  argTypes: {
    children: { type: null }
  }
};

// used to center the tooltip button/icon in the viewport
const CenterContent = ({ children }) => (
  <div
    style={{
      position: 'absolute',
      top: 0,
      bottom: 0,
      width: '100%',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center'
    }}
  >
    {children}
  </div>
);

const ButtonTemplate = (args) => (
  <CenterContent>
    <Tooltip {...args}>
      <Button>Mouse over or focus on me</Button>
    </Tooltip>
  </CenterContent>
);

const defaultArgs = {
  content: 'Example tooltip content'
};

export const top = ButtonTemplate.bind({});
top.args = {
  ...defaultArgs,
  position: 'top'
};

export const left = ButtonTemplate.bind({});
left.args = {
  ...defaultArgs,
  position: 'left'
};

export const bottom = ButtonTemplate.bind({});
bottom.args = {
  ...defaultArgs,
  position: 'bottom'
};

export const right = ButtonTemplate.bind({});
right.args = {
  ...defaultArgs
};

const IconTemplate = (args) => (
  <CenterContent>
    <div>
      <p>Try mousing over or focusing on this icon:</p>

      <Tooltip {...args}>
        <div>
          <svg style={{ width: 20, height: 20 }} viewBox="0 0 512 512">
            <g>
              <path
                d={`	
                M256 6C118 6 6 118 6 256s112 250 250	
                250 250-112 250-250S394 6 256 6zm-40.3	
                76.6h84.7V151h-84.7V82.7zm121	
                346.8H175.3V361h32.3v-97h-32.3v-68.4h121V361h40.3v68.4z	
              `}
                fill="#000"
              />
            </g>
          </svg>
        </div>
      </Tooltip>
    </div>
  </CenterContent>
);

export const icon = IconTemplate.bind({});
icon.args = {
  ...defaultArgs
};
