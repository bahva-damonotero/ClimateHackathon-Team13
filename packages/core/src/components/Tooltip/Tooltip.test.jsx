import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import tooltip from '@uswds/uswds/js/usa-tooltip';

import Tooltip from '.';
import { Button } from '..';

describe('Tooltip', () => {
  const content = 'Tooltip content';

  const jsx = (
    <Tooltip content={content}>
      <Button>Tooltip example</Button>
    </Tooltip>
  );

  it('should render correctly', () => {
    const { container } = render(jsx);

    const tooltipNode = container.querySelector(
      '.bah-uswds-tooltip .usa-tooltip'
    );

    expect(
      tooltipNode.querySelector('.usa-tooltip__trigger').dataset.position
    ).toBe('right');

    expect(tooltipNode.querySelector('.usa-tooltip__body').innerHTML).toBe(
      content
    );
  });

  it('should call `tooltip.on()` to initialize the component', () => {
    const onSpy = jest.spyOn(tooltip, 'on');

    render(jsx);

    expect(onSpy).toHaveBeenCalled();
  });

  it('should match snapshot', () => {
    const tree = renderer.create(jsx).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
