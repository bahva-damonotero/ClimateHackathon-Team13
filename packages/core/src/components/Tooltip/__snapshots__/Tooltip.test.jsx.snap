// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should match snapshot 1`] = `
<div
  className="bah-uswds-tooltip"
>
  <button
    className="usa-button usa-tooltip"
    data-position="right"
    disabled={false}
    onClick={[Function]}
    title="Tooltip content"
    type="button"
  >
    Tooltip example
  </button>
</div>
`;
